README.md
# 3scale Istio Mixer gRPC Adapter

An [out of process gRPC Adapter](https://github.com/istio/istio/wiki/Mixer-Out-Of-Process-Adapter-Dev-Guide) which integrates 3scale with Istio

* [Overview](#overview)
* [Prerequisites](#prerequisites)
* [Enabling Policies](#enabling-policies)
* [Deploy the adapter](#deploy-the-adapter)
* [Create the required resources](#create-the-required-resources)
* [Generating and creating configuration](#generating-and-creating-configuration)
* [Routing service traffic through the adapter](#routing-service-traffic-through-the-adapter)
* [Authenticating requests](#authenticating-requests)
  * [Applying Patterns](#applying-patterns)
    * [API Key Pattern](#api-key-pattern)
    * [Application ID Pattern](#application-id-pattern)
    * [OpenID Connect Pattern](#openid-connect-pattern)
    * [Hybrid](#hybrid-pattern)
* [Adapter metrics](#adapter-metrics)
* [Development and contributing](#development-and-contributing)

## Overview

When running Istio in a Kubernetes or OpenShift cluster, this adapter allows a user to label their service
running within the mesh, and have that service integrated with the [3scale Api Management solution](https://www.3scale.net/).

## Prerequisites

1. Istio version 1.1 with [policies enabled](#enabling-policies)
1. A working [3scale account](https://www.3scale.net/signup) (SaaS or On-Premises)
1. `kubectl` or `oc` command line tool

## Enabling Policies

As of Istio 1.1.0 GA, policies are now disabled by default. This impacts the 3scale adapter and policies need to be re-enabled
in order for the adapter to receive traffic.

Follow [these instructions](https://istio.io/docs/tasks/policy-enforcement/enabling-policy/) to enable policies.

## Deploy the adapter

A Kubernetes deployment and service manifest are located in the `deploy` directory.
To deploy the adapter to a Kubernetes/OpenShift cluster, run:

```bash
kubectl create -f deploy/
```

## Configuring the adapter

See [the adapter configuration options](cmd/server/README.md) to understand the default behaviour of the adapter, and how to modify it.

## Create the required resources

The required CustomResources are located in the `istio` directory.

Create the required resources in the desired istio namespace. By default, this is `istio-system`, however, in a multi-tenant environment,
this could be different.

```bash
kubectl create -f istio/authorization-template.yaml
kubectl creae -f istio/threescale-adapter.yaml
```

## Generating and creating configuration

The adapter embeds a tool which allows generation of the `handler`,`instance` and `rule` CR's.
More detail can be found in the tools [documentation](cmd/cli/README.md).

To generate these manifests from a deployed adapter, assuming it is deployed in the `istio-system` namespace, run the following:

```bash
export NS="istio-system" URL="https://replaceme-admin.3scale.net:443" NAME="name" TOKEN="token"
oc exec -n ${NS} $(oc get po -n ${NS} -o jsonpath='{.items[?(@.metadata.labels.app=="3scale-istio-adapter")].metadata.name}') \
-it -- ./3scale-config-gen \
--url ${URL} --name ${NAME} --token ${TOKEN} -n ${NS}
```

This will produce some sample output to the terminal. Edit these samples if required and create the objects using `oc create` command.


Update the workload (target service deployment's Pod Spec) with the required annotations:

```bash
export CREDENTIALS_NAME="replace-me"
export SERVICE_ID="replace-me"
export DEPLOYMENT="replace-me"
patch="$(oc get deployment "${DEPLOYMENT}" --template='{"spec":{"template":{"metadata":{"labels":{ {{ range $k,$v := .spec.template.metadata.labels }}"{{ $k }}":"{{ $v }}",{{ end }}"service-mesh.3scale.net/service-id":"'"${SERVICE_ID}"'","service-mesh.3scale.net/credentials":"'"${CREDENTIALS_NAME}"'"}}}}}' )"
oc patch deployment "${DEPLOYMENT}" --patch ''"${patch}"''
```

Sample manifest can be found in the `testdata` directory which can be used as a base if you prefer not to leverage the tool.

## Routing service traffic through the adapter

The following statements assume that configuration has been generated using the tool provided. Custom modifications
to the match condition in the rule are possible, but explanation of any or all combinations is outside of the scope of
this document. If interested, check out the [upstream documentation](https://istio.io) 

In order to drive traffic for your service through the adapter and be managed by 3scale, we need to match the rule
`destination.labels["service-mesh.3scale.net/credentials"] == "threescale"` we previously created in the configuration, in the `kind: rule` resource.

Integration of a service requires that the above label be added to PodTemplateSpec on the Deployment of the target workload.
The value, `threescale`, refers to the name of the generated handler. This handler will store the access token required to call 3scale.

When the request reaches the adapter, the adapter needs to know the how the service maps to the an API on 3scale.
This can be provided in one of two ways:

1. As a label on the workload (recommended)
1. Hardcoded in the handler as `service_id`

To pass the service ID to the adapter via the instance at request time, add the following label to the workload:

`destination.labels["service-mesh.3scale.net/service-id"] == "replace-me"`

Your 3scale administrator should be able to provide you with both the required credentials name and the service ID.

## Authenticating requests

Now that the we have [configured the service to be managed by 3scale](#routing-service-traffic-through-the-adapter) we can decide how requests should be authenticated.
Currently there are two supported mechanisms:
1. The API Key authentication pattern
2. The Application ID, Application Key (optional) pair authentication pattern

You can read more detailed information about these patterns and their behaviour in the [3scale documentation](https://access.redhat.com/documentation/en-us/red_hat_3scale_api_management/2.4/html/api_authentication/authentication-patterns).


### Applying Patterns

When you have decided what pattern best fits your needs, you can modify the `instance` CustomResource to configure this behaviour.
You can also decide if authentication credentials should be read from headers or query parameters, or allow both.

It is important to note that when specifying values from headers, Istio expects they key to be lower case.
So for example if you want to send a header as `User-Key`, this must be referenced in the configuration as `request.headers["user-key"]`.


#### API Key Pattern
To use the *API Key authentication pattern*, you should use the `user` value on the `subject` field like so:

```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: instance
metadata:
  name: threescale-authorization
spec:
  template: threescale-authorization
  params:
    subject:
      user: request.query_params["user_key"] | request.headers["user_key"] | ""
    action:
      path: request.url_path
      method: request.method | "get"
      service: destination.labels["service-mesh.3scale.net/service-id"] | ""
```

This configuration will examine the `user_key` query parameter, followed by the `user-key` header in search of the api key. As mentioned, this can be restricted to one or the other by removing that particular attribute.
The order can be changed to determine precedence.

If you would like for the adapter to examine a different, for example query parameter than `user_key`, you would simply change `[user_key]` to `[foo]`. The same pattern applies to the headers.

#### Application ID Pattern
To use the *Application ID authentication pattern*, you should use the `properties` value on the `subject` field to set `app_id`, and **optionally** `app_key`.

Manipulation of this object can be done in using the methods described previously.
An example configuration is shown below.

```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: instance
metadata:
  name: threescale-authorization
spec:
  template: threescale-authorization
  params:
    subject:
        app_id: request.query_params["app_id"] | request.headers["app_id"] | ""
        app_key: request.query_params["app_key"] | request.headers["app_key"] | ""
    action:
      path: request.url_path
      method: request.method | "get"
      service: destination.labels["service-mesh.3scale.net/service-id"] | ""
```

#### OpenID Connect Pattern
To use the *OpenID Connect authentication pattern*, you should use the `properties` value on the `subject` field to set `client_id`, and **optionally** `app_key`.

Manipulation of this object can be done in using the methods described previously.
An example configuration is shown below. Here the client identifier (application ID) is parsed from the JWT under the label "azp". Modify this as desired.

```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: instance
metadata:
  name: threescale-authorization 
spec:
  template: threescale-authorization
  params:
    subject:
        app_key: request.query_params["app_key"] | request.headers["app_key"] | ""
        client_id: request.auth.claims["azp"] | ""
    action:
      path: request.url_path
      method: request.method | "get"
      service: destination.labels["service-mesh.3scale.net/service-id"] | ""
```

For this integration to work correctly. OpenID configuration must still be done in 3scale in order for the client to be created in the IdP.
For the service the user wants to protect, they should create [Request authorization](https://istio.io/latest/docs/tasks/security/authorization/authz-jwt//) in the same namespace as that service.
The JWT should then be passed in the `Authorization` header of the request.

In the sample `RequestAuthentication` defined below, replace `issuer`, `jwksUri` and `selector` as appropriate.

```yaml
- apiVersion: security.istio.io/v1beta1
  kind: RequestAuthentication
  metadata:
    name: jwt-example
    namespace: bookinfo
  spec:
    selector:
      matchLabels:
        app: productpage
    jwtRules:
    - issuer: >-
        http://keycloak-keycloak.34.242.107.254.nip.io/auth/realms/3scale-keycloak
      jwksUri: >-
        http://keycloak-keycloak.34.242.107.254.nip.io/auth/realms/3scale-keycloak/protocol/openid-connect/certs
```

#### Hybrid Pattern

Finally, you may decide to not enforce a particular authentication method but accept any valid credentials for either pattern. In that case, you can do a hybrid configuration where the user key pattern will be preferred if both are provided:
```yaml
apiVersion: "config.istio.io/v1alpha2"
kind: instance
metadata:
  name: threescale-authorization
spec:
  template: threescale-authorization
  params:
    subject:
      user: request.query_params["user_key"] | request.headers["user_key"] | ""
      properties:
        app_id: request.query_params["app_id"] | request.headers["app_id"] | ""
        app_key: request.query_params["app_key"] | request.headers["app_key"] | ""
    action:
      path: request.url_path
      method: request.method | "get"
      service: destination.labels["service-mesh.3scale.net/service-id"] | ""

```

## Adapter metrics

The adapter, by default reports various Prometheus metrics which are exposed on port `8080` at the `/metrics` endpoint.
These allow some insight into how the interactions between the adapter and 3scale are performing. The service gets labelled
and automatically discovered and scraped by Prometheus.


## Development and contributing

Check the [DEVEL.md](DEVEL.md) for more info on how to hack/test this adapter.


CHANGELOG.md
# Change Log

Notable changes to 3scale Istio Mixer Adapter will be tracked in this document.

## 2.0.3 - 2021-06-14

### Added

- Added CI tests for the ppc64le architecture. ([#185](https://github.com/3scale/3scale-istio-adapter/pull/185))

### Fixed

- No longer require for OIDC flows to provide the application's `client_secret`
  as the `app_key` parameter. ([#188](https://github.com/3scale/3scale-istio-adapter/pull/188))
- Go 1.16 compilation. ([#187](https://github.com/3scale/3scale-istio-adapter/pull/187))
- Fixed a few minor warnings in the code. ([#186](https://github.com/3scale/3scale-istio-adapter/pull/186))

## 2.0.2 - 2021-03-02

### Added

- Support for specifying a set of root CAs via a PEM file (in addition to system CAs)
  using the `ROOT_CA` environment variable. ([#181](https://github.com/3scale/3scale-istio-adapter/pull/181))
- Support for specifying a client certificate and key via PEM files using the
  `CLIENT_CERT` and `CLIENT_KEY` environment variables. ([#181](https://github.com/3scale/3scale-istio-adapter/pull/181))

## 2.0.1.1 - 2021-02-04

### Fixed

- The sample custom resources for JWT authentication in the README have been
  updated from Policy to RequestAuthentication. ([#177](https://github.com/3scale/3scale-istio-adapter/pull/177))
- Examples for 3scale-config-gen in documentation where incorrectly spelled as
  3scale-gen-config. ([#178](https://github.com/3scale/3scale-istio-adapter/pull/178))

## 2.0.1 - 2020-09-29

### Fixed

- The configuration for the failure policy (`BACKEND_CACHE_POLICY_FAIL_CLOSED`)
  was not being taken into account and its default value was false rather than
  true. ([#175](https://github.com/3scale/3scale-istio-adapter/pull/175))
- Added missing documentation in the gRPC server for the flush intervals and
  the failure policy, and removed the unused `THREESCALE_` prefix. ([#175](https://github.com/3scale/3scale-istio-adapter/pull/175))


## 2.0.0.1 - 2020-07-30

### Security

- Updated dependencies on golang.org/x/text and golang.org/x/crypto to fix CVE-2020-14040 and CVE-2020-9283.
  ([#171](https://github.com/3scale/3scale-istio-adapter/pull/171))

## 2.0.0 - 2020-07-24

### Added

- An authorization cache which maintains local counters for rate limits and
  periodically flushes metrics to 3scale to increase performance.
  ([#167](https://github.com/3scale/3scale-istio-adapter/pull/167))
- A ConfigMap to provide configuration to the gRPC adapter
  ([#164](https://github.com/3scale/3scale-istio-adapter/pull/164))
    * A configuration option to the gRPC server (`USE_CACHED_BACKEND`) to enable the authorization cache.
    * A configuration option to the gRPC server (`BACKEND_CACHE_FLUSH_INTERVAL_SECONDS`) to set the interval at which metrics get reported from the cache to 3scale.
    * A configuration option to the gRPC server (`BACKEND_CACHE_POLICY_FAIL_CLOSED`) to determine the fate of a request if 3scale Apisonator is unreachable.
- Prometheus metrics for authorization cache
  ([#162](https://github.com/3scale/3scale-istio-adapter/pull/162))
- Support for `last` and priority of mapping rules as defined in 3scale Porta
  ([#150](https://github.com/3scale/3scale-istio-adapter/pull/150))

### Changed

- Removal of `THREESCALE_` prefix from existing environment variables
  ([#164](https://github.com/3scale/3scale-istio-adapter/pull/164))
- Metrics have been converged to include both Porta and Apisonator requests, differentiated by labels.
  ([#162](https://github.com/3scale/3scale-istio-adapter/pull/162))
- Updated response codes to match HTTP equivalent, as returned by APIcast
  ([#144](https://github.com/3scale/3scale-istio-adapter/pull/144))
  ([#138](https://github.com/3scale/3scale-istio-adapter/pull/138))
  ([#136](https://github.com/3scale/3scale-istio-adapter/pull/136))

### Fixed

- Issue with `help` command for CLI tool returning excess information from imported dependency.
  ([#151](https://github.com/3scale/3scale-istio-adapter/pull/151))

## 1.0.0.1 - 2020-07-29

### Security

- Updated dependencies on golang.org/x/text and golang.org/x/crypto to fix CVE-2020-14040 and CVE-2020-9283.
  ([#171](https://github.com/3scale/3scale-istio-adapter/pull/171))

## 1.0.0 - 2019-08-07

### Added

- A field to the handler params (`backend_url`) to allow overriding the 3scale backend the adapter should reach out to.
  ([#111](https://github.com/3scale/3scale-istio-adapter/pull/111))

### Fixed

- The CLI tool no longer panics when provided with a name that fails Kubernetes validation.
  ([#113](https://github.com/3scale/3scale-istio-adapter/pull/113))

## 0.7.1 - 2019-06-17

### Added

- A configuration option to the gRPC server (`THREESCALE_LOG_GRPC`) to allow suppression of gRPC logging.
  A configuration option to the gRPC server (`THREESCALE_GRPC_CONN_MAX_SECONDS`) to allow setting specific keepalive parameters.
  ([#104](https://github.com/3scale/3scale-istio-adapter/pull/104))

### Changed

- The Kubernetes service is now headless to support client side load balancing.
  The CLI tool now generates the connection address prefixed with `dns:///`.
  ([#104](https://github.com/3scale/3scale-istio-adapter/pull/104))
- The dependency on Istio's `api` and `istio` packages is now based on version `1.1.8`
  ([#108](https://github.com/3scale/3scale-istio-adapter/pull/108))


## 0.7.0 - 2019-06-12

### Added

- Support for reading a services 3scale service ID from the pod's label providing a service discovery mechanism.
  This in turn requires a `name` flag to be added to the CLI tool.
  ([#93](https://github.com/3scale/3scale-istio-adapter/pull/93))
- A `namesapce` flag has been added to the CLI tool to support multi-tenancy in Maistra.
  ([#103](https://github.com/3scale/3scale-istio-adapter/pull/103))

### Changed

- The CLI tool now generates different output:
  The `destination.labels["service-mesh.3scale.net"]` and `destination.labels["service-mesh.3scale.net/uid"]` labels have been changed solely to

scripts/deployment.go
package main

/*
 Utility script to generate deployment template prior to release
*/

import (
	"fmt"
	"log"

	"github.com/ghodss/yaml"
	v1 "k8s.io/api/apps/v1"
	corev1 "k8s.io/api/core/v1"
	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
	"k8s.io/apimachinery/pkg/util/intstr"
)

var version string

const resource = "Deployment"
const resourceName = "3scale-istio-adapter"
const selectorKey = "app"
const selectorValue = resourceName

const adapterImageRegistry = "quay.io/3scale/"
const adapterImage = resourceName
const adapterPort = 3333
const metricPort = 8080

const configMapName = "3scale-istio-adapter-conf"

const deploymentStrategy = "RollingUpdate"

var replicas int32 = 1
var terminationGraceSeconds int64 = 30

func main() {
	if version == "" {
		log.Fatal("version must be set")
	}
	now := metav1.Now()

	deployment := v1.Deployment{
		TypeMeta: metav1.TypeMeta{
			Kind:       resource,
			APIVersion: "apps/v1",
		},
		ObjectMeta: metav1.ObjectMeta{
			Name: resourceName,
			Labels: map[string]string{
				selectorKey: selectorValue,
			},
			CreationTimestamp: now,
		},
		Spec: v1.DeploymentSpec{
			Replicas: &replicas,
			Selector: &metav1.LabelSelector{
				MatchLabels: map[string]string{
					selectorKey: selectorValue,
				},
			},
			Strategy: v1.DeploymentStrategy{
				Type: v1.DeploymentStrategyType(deploymentStrategy),
				RollingUpdate: &v1.RollingUpdateDeployment{
					MaxUnavailable: &intstr.IntOrString{
						IntVal: 25,
					},
					MaxSurge: &intstr.IntOrString{
						IntVal: 25,
					},
				},
			},
			Template: corev1.PodTemplateSpec{
				ObjectMeta: metav1.ObjectMeta{
					Labels: map[string]string{
						selectorKey: selectorValue,
					},
					CreationTimestamp: now,
				},
				Spec: corev1.PodSpec{
					Containers: []corev1.Container{
						{
							Name:            resourceName,
							Image:           adapterImageRegistry + adapterImage + ":" + version,
							ImagePullPolicy: corev1.PullAlways,
							Ports: []corev1.ContainerPort{
								{
									Name:          "adapter",
									ContainerPort: adapterPort,
									Protocol:      corev1.ProtocolTCP,
								},
								{
									Name:          "prometheus",
									ContainerPort: metricPort,
									Protocol:      corev1.ProtocolTCP,
								},
							},
							LivenessProbe: &corev1.Probe{
								Handler: corev1.Handler{
									TCPSocket: &corev1.TCPSocketAction{
										Port: intstr.IntOrString{
											IntVal: adapterPort,
										},
									},
								},
								InitialDelaySeconds: 10,
								PeriodSeconds:       5,
							},
							Env: []corev1.EnvVar{
								{
									Name: "LOG_LEVEL",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "log_level",
										},
									},
								},
								{
									Name: "LOG_JSON",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "log_json",
										},
									},
								},
								{
									Name: "REPORT_METRICS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "metrics.report",
										},
									},
								},
								{
									Name: "CACHE_TTL_SECONDS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "system.cache_ttl",
										},
									},
								},
								{
									Name: "CACHE_REFRESH_SECONDS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "system.cache_ttl",
										},
									},
								},
								{
									Name: "CACHE_ENTRIES_MAX",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "system.cache_max_size",
										},
									},
								},
								{
									Name: "CACHE_REFRESH_RETRIES",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "system.cache_refresh_retries",
										},
									},
								},
								{
									Name: "ALLOW_INSECURE_CONN",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "client.allow_insecure_connections",
										},
									},
								},
								{
									Name: "ROOT_CA",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "client.root_ca",
										},
									},
								},
								{
									Name: "CLIENT_CERT",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "client.client_cert",
										},
									},
								},
								{
									Name: "CLIENT_KEY",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "client.client_key",
										},
									},
								},
								{
									Name: "CLIENT_TIMEOUT_SECONDS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "client.timeout",
										},
									},
								},
								{
									Name: "GRPC_CONN_MAX_SECONDS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "grpc.max_conn_timeout",
										},
									},
								},
								{
									Name: "USE_CACHED_BACKEND",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "backend.enable_cache",
										},
									},
								},
								{
									Name: "BACKEND_CACHE_FLUSH_INTERVAL_SECONDS",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "backend.cache_flush_interval",
										},
									},
								},
								{
									Name: "BACKEND_CACHE_POLICY_FAIL_CLOSED",
									ValueFrom: &corev1.EnvVarSource{
										ConfigMapKeyRef: &corev1.ConfigMapKeySelector{
											LocalObjectReference: corev1.LocalObjectReference{
												Name: configMapName,
											},
											Key: "backend.policy_fail_closed",
										},
									},
								},
							},
							Resources:              corev1.ResourceRequirements{},
							TerminationMessagePath: "/dev/termination-log",
						},
					},
					DNSPolicy:                     corev1.DNSClusterFirst,
					RestartPolicy:                 corev1.RestartPolicyAlways,
					SecurityContext:               &corev1.PodSecurityContext{},
					TerminationGracePeriodSeconds: &terminationGraceSeconds,
				},
			},
		},
	}

	b, err := yaml.Marshal(deployment)
	if err != nil {
		panic("unexpected marshal error" + err.Error())
	}
	fmt.Printf("# This code was generated as part of the release process using make release for version %s\n%s", version, string(b))
}
